File: parse_lib/IPPcode21.php
<?php

interface IPPCode
{
    // Frame and Function Calls Instructions
    const HEADER        = ".IPPcode21";
    const MOVE          = "MOVE";
    const CREATEFRAME   = "CREATEFRAME";
    const PUSHFRAME     = "PUSHFRAME";
    const POPFRAME      = "POPFRAME";
    const DEFVAR        = "DEFVAR";
    const CALL          = "CALL";
    const RETURN        = "RETURN";

    // Data Stack Instructions
    const PUSHS         = "PUSHS";
    const POPS          = "POPS";

    // Arithmetic, Relation, Boolean, and Conversion Intstructions
    const ADD           = "ADD";
    const SUB           = "SUB";
    const MUL           = "MULL";
    const IDIV          = "IDIV";
    const LT            = "LT";
    const GT            = "GT";
    const EQ            = "EQ";
    const AND           = "AND";
    const OR            = "OR";
    const NOT           = "NOT";
    const INT2CHAR      = "INT2CHAR";
    const STRI2INT      = "STRI2INT";

    // Input, Output Instructions
    const READ          = "READ";
    const WRITE         = "WRITE";

    // String Instructions
    const CONCAT        = "CONCAT";
    const STRLEN        = "STRLEN";
    const GETCHAR       = "GETCHAR";
    const SETCHAR       = "SETCHAR";

    // Type Instructions
    const TYPE          = "TYPE";

    // Flow Instructions
    const LABEL         = "LABEL";
    const JUMP          = "JUMP";
    const JUMPIFEQ      = "JUMPIFEQ";
    const JUMPIFNEQ     = "JUMPIFNEQ";
    const EXIT          = "EXIT";

    // Debugging Instructions
    const DPRINT        = "DPRINT";
    const BREAK         = "BREAK";
}

File: interpret_lib/ippcode21.py
class IppCode21:
    instructions = [
        "MOVE",
        "CREATEFRAME",
        "PUSHFRAME",
        "POPFRAME",
        "DEFVAR",
        "CALL",
        "RETURN",
        # Data Stack Instructions
        "PUSHS",
        "POPS",
        # Arithmetic, Relation, Boolean, and Conversion Intstructions
        "ADD",
        "SUB",
        "MUL",
        "IDIV",
        "LT",
        "GT",
        "EQ",
        "AND",
        "OR",
        "NOT",
        "INT2CHAR",
        "STRI2INT",
        # Input, Output Instructions
        "READ",
        "WRITE",
        # String Instructions
        "CONCAT",
        "STRLEN",
        "GETCHAR",
        "SETCHAR",
        # Type Instructions
        "TYPE",
        # Flow Instructions
        "LABEL",
        "JUMP",
        "JUMPIFEQ",
        "JUMPIFNEQ",
        "EXIT",
        # Debugging Instructions
        "DPRINT",
        "BREAK",
        # Float extension
        "INT2FLOAT",
        "FLOAT2INT",
        "DIV",
    ]

    stack_instructions = [
        "CLEARS",
        "ADDS",
        "SUBS",
        "MULS",
        "IDIVS",
        "LTS",
        "GTS",
        "EQS",
        "ANDS",
        "ORS",
        "NOTS",
        "INT2CHARS",
        "STRI2INTS",
        "JUMPIFEQS",
        "JUMPIFNEQS",
    ]
